packages/core/src/factory.ts

```diff
@@ -18,62 +22,71 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
     request: (input: Input, resources: Resource[], c?: Config) => {
       // build the config for this request
       const config = defaults<unknown, Config>(c, defaultConfig);
-      const inputValidations: IPrismDiagnostic[] = [];
 
       return pipe(
         TaskEither.fromEither(components.route({ resources, input })),
-        TaskEither.chain(resource => {
-          if (config.validateRequest) {
-            inputValidations.push(
-              ...components.validateInput({
-                resource,
-                element: input,
-              })
-            );
-          }
-
-          config.checkSecurity
-            ? inputValidations.push(
-                ...pipe(
-                  validateSecurity(input, resource),
-                  map(sec => [sec]),
-                  getOrElse<IPrismDiagnostic[]>(() => [])
-                )
+        TaskEither.chain(resource =>
+          TaskEither.fromEither(
+            pipe(
+              sequenceValidation(
+                config.validateRequest ? components.validateInput({ resource, element: input }) : Either.right(input),
+                config.checkSecurity ? validateSecurity(input, resource) : Either.right(input)
+              ),
+              Either.fold(
+                validations =>
+                  Either.right<Error, { resource: Resource; validations: IPrismDiagnostic[] }>({
+                    resource,
+                    validations,
+                  }),
+                () =>
+                  Either.right<Error, { resource: Resource; validations: IPrismDiagnostic[] }>({
+                    resource,
+                    validations: [],
+                  })
```

Comment:
note to self: so `Either.right` because we also want to gather output validation results (?)

Reply:
Fundamentally no matter if the validation fails or succeeds — we still want to call the mock method with empty validation array or not — and the only way to make the chain continue is to make sure the Either returned is a right.

The successive step would be to break down the `mock` method into the `mockSuccess` and `mockFailure` (or something similar) so that we can properly use the fp-ts flows.

I still do not have a clear plan about it though, but this is the direction we should be taking.